File: packages/schema/src/migrations/v0_6_to_v0_7.ts
/**
 * Migration from schema version 0.6.0 to 0.7.0.
 *
 * Renames StorySegmentLayer content.title → imageCaption.
 */

export function migrate(doc: Record<string, any>): Record<string, any> {
  const layers: Record<string, any> = { ...doc.layers };

  for (const [id, layer] of Object.entries(layers)) {
    if (layer?.type !== 'StorySegmentLayer') {
      continue;
    }

    const parameters = layer.parameters;
    const content = parameters?.content;

    if (
      !content ||
      typeof content !== 'object' ||
      !('title' in content) ||
      content.imageCaption !== undefined
    ) {
      continue;
    }

    const { title, ...rest } = content;

    layers[id] = {
      ...layer,
      parameters: {
        ...parameters,
        content: {
          ...rest,
          imageCaption: title ?? '',
        },
      },
    };
  }

  return { ...doc, layers };
}

File: python/jupytergis_core/jupytergis_core/migrations/v0_6_to_v0_7.py
"""Migration from schema version 0.6.0 to 0.7.0.

Renames StorySegmentLayer content.title → imageCaption.
"""

from typing import Any


def migrate(doc: dict[str, Any]) -> dict[str, Any]:
    layers = dict(doc.get("layers", {}))

    for layer_id, layer in layers.items():
        if layer.get("type") != "StorySegmentLayer":
            continue

        parameters = layer.get("parameters") or {}
        content = parameters.get("content")

        if (
            not isinstance(content, dict)
            or "title" not in content
            or content.get("imageCaption") is not None
        ):
            continue

        rest = {k: v for k, v in content.items() if k != "title"}
        layers[layer_id] = {
            **layer,
            "parameters": {
                **parameters,
                "content": {
                    **rest,
                    "imageCaption": content.get("title") or "",
                },
            },
        }

    return {**doc, "layers": layers}
